### ISSUE TITLE
cue/parser: ParseFile should read from filename when src is a typed nil

### ISSUE THREAD

**Issue by asdine** [cue-lang/cue#1929]
### What version of CUE are you using (`cue version`)?

<pre>
$ cue version
cue version v0.4.3 darwin/arm64
</pre>

### Does this issue reproduce with the latest release?

Yes

### What did you do?

[parser.ParseFile](https://pkg.go.dev/cuelang.org/go@v0.4.3/cue/parser#ParseFile) returns silently when passing a nil byte slice instead of reading the file from disk.

```go
// this works
content, err := source.Read("foo.cue", nil)

// this doesn't
func bar(filename string, content []byte) {
  content, err := source.Read(filename, content)
}
```

```
-- foo.cue --
a: 1
-- go.mod --
module foo

go 1.19

require cuelang.org/go v0.4.3

require (
	github.com/cockroachdb/apd/v2 v2.0.1 // indirect
	github.com/mpvl/unique v0.0.0-20150818121801-cbe035fff7de // indirect
	github.com/pkg/errors v0.8.1 // indirect
)
-- go.sum --
cuelang.org/go v0.4.3 h1:W3oBBjDTm7+IZfCKZAmC8uDG0eYfJL4Pp/xbbCMKaVo=
cuelang.org/go v0.4.3/go.mod h1:7805vR9H+VoBNdWFdI7jyDR3QLUPp4+naHfbcgp55HI=
github.com/cockroachdb/apd v1.1.0 h1:3LFP3629v+1aKXU5Q37mxmRxX/pIu1nijXydLShEq5I=
github.com/cockroachdb/apd/v2 v2.0.1 h1:y1Rh3tEU89D+7Tgbw+lp52T6p/GJLpDmNvr10UWqLTE=
github.com/cockroachdb/apd/v2 v2.0.1/go.mod h1:DDxRlzC2lo3/vSlmSoS7JkqbbrARPuFOGr0B9pvN3Gw=
github.com/davecgh/go-spew v1.1.1 h1:vj9j/u1bqnvCEfJOwUhtlOARqs3+rkHYY13jYWTU97c=
github.com/emicklei/proto v1.6.15 h1:XbpwxmuOPrdES97FrSfpyy67SSCV/wBIKXqgJzh6hNw=
github.com/golang/glog v0.0.0-20160126235308-23def4e6c14b h1:VKtxabqXZkF25pY9ekfRL6a582T4P37/31XEstQ5p58=
github.com/google/go-cmp v0.4.0 h1:xsAVV57WRhGj6kEIi8ReJzQlHHqcBYCElAvkovg3B/4=
github.com/google/uuid v1.2.0 h1:qJYtXnJRWmpe7m/3XlyhrsLrEURqHRM2kxzoxXqyUDs=
github.com/lib/pq v1.0.0 h1:X5PMW56eZitiTeO7tKzZxFCSpbFZJtkMMooicw2us9A=
github.com/mpvl/unique v0.0.0-20150818121801-cbe035fff7de h1:D5x39vF5KCwKQaw+OC9ZPiLVHXz3UFw2+psEX+gYcto=
github.com/mpvl/unique v0.0.0-20150818121801-cbe035fff7de/go.mod h1:kJun4WP5gFuHZgRjZUWWuH1DTxCtxbHDOIJsudS8jzY=
github.com/pkg/errors v0.8.0/go.mod h1:bwawxfHBFNV+L2hUp1rHADufV3IMtnDRdf1r5NINEl0=
github.com/pkg/errors v0.8.1 h1:iURUrRGxPUNPdy5/HRSm+Yj6okJ6UtLINN0Q9M4+h3I=
github.com/pkg/errors v0.8.1/go.mod h1:bwawxfHBFNV+L2hUp1rHADufV3IMtnDRdf1r5NINEl0=
github.com/pmezard/go-difflib v1.0.0 h1:4DBwDE0NGyQoBHbLQYPwSUPoCMWR5BEzIk/f1lZbAQM=
github.com/protocolbuffers/txtpbfmt v0.0.0-20201118171849-f6a6b3f636fc h1:gSVONBi2HWMFXCa9jFdYvYk7IwW/mTLxWOF7rXS4LO0=
github.com/rogpeppe/go-internal v1.8.1 h1:geMPLpDpQOgVyCg5z5GoRwLHepNdb71NXb67XFkP+Eg=
github.com/stretchr/testify v1.2.2 h1:bSDNvY7ZPG5RlJ8otE/7V6gMiyenm9RtJ7IUVIAoJ1w=
golang.org/x/net v0.0.0-20200226121028-0de0cce0169b h1:0mm1VjtFUOIlE1SbDlwjYaDxZVDP2S5ou6y0gSgXHu8=
golang.org/x/text v0.3.7 h1:olpwvP2KacW1ZWvsR7uQhoyTYvKAupfQrRGBFM352Gk=
golang.org/x/xerrors v0.0.0-20191204190536-9bdfabe68543 h1:E7g+9GITq07hpfrRu66IVDexMakfv52eLZ2CXBWiKr4=
gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b h1:h8qDotaEPuJATrMmW04NCwg7v22aHH28wwpauUhK9Oo=
-- main.go --
package main

import (
	"fmt"
	"strconv"

	"cuelang.org/go/cue/ast"
	"cuelang.org/go/cue/parser"
)

func main() {
	a, err := getA("foo.cue", nil)
	if err != nil {
		panic(err)
	}
	fmt.Println(a)
}

func getA(filename string, content []byte) (int, error) {
	f, err := parser.ParseFile(filename, content)
	if err != nil {
		return 0, err
	}

	for _, d := range f.Decls {
		field, ok := d.(*ast.Field)
		if !ok {
			continue
		}
		if field.Label.(*ast.Ident).Name == "a" {
			return strconv.Atoi(field.Value.(*ast.BasicLit).Value)
		}
	}

	return -1, nil
}
-- saved.txtar --
```

### What did you expect to see?

`1`

Basically, parser.ParseFile should detect if the slice is actually empty.

### What did you see instead?

`-1`

Happy to submit a PR to fix it.

**Comment by mvdan** [comment]
This is a tricky one; I'm willing to say that this is working as intended. We only read from the filename if src is an _untyped nil_, whereas you are passing a _typed nil_ in the form of `[]byte(nil)`, which is a valid way of specifying zero bytes.

This is such a common "gotcha" in Go that it's even in its FAQ: https://go.dev/doc/faq#nil_error

Also note that our current behavior is the same that upstream Go has in its parser package: https://cs.opensource.google/go/go/+/refs/tags/go1.19.1:src/go/parser/interface.go;drc=c7a0b156592ca15612315fc71f4e287268643bfd;l=24

I am wondering if we can improve the docs, though they do clearly say:

> If src == nil, ParseFile parses the file specified by filename.

Also worth noting that changing this behavior would potentially break some users.

**Comment by mvdan** [comment]
Another way to look at this is that, if we swap the behavior, then parsing an empty file will be confusing. Both `[]byte(nil)` and `[]byte{}` are empty Go byte slices, but only the first is an untyped nil (`== nil`), so the first will actually read from the filename on disk rather than parsing the passed bytes.

Also worth noting that it's common for Go APIs to treat a `nil` slice equally to an empty but non-nil slice to avoid confusion and surprises. I can't find a piece of docs on this right now, but you should see that none of the APIs in the `bytes` package special case a nil input, for example.

**Comment by asdine** [comment]
I understand, since this mimics the Go API and that any change might break some users, I guess adding a comment would be great.
FWIW, I still think that this API creates confusion, as a lot of APIs treat `nil`, `[]T(nil)` and `[]T{}` as the same thing. I got tricked by it twice in the same week (I'm not the most observant person though).
But I don't see a better way to rewrite it, so I guess we should keep it as is.

**Comment by mvdan** [comment]
> FWIW, I still think that this API creates confusion, as a lot of APIs treat `nil`, `[]T(nil)` and `[]T{}` as the same thing.

Right, except that in this API, due to the interface type, `nil` and `[]T(nil)` are very different values :) For APIs that take a slice type as a parameter, the two values are exactly the same.

> But I don't see a better way to rewrite it, so I guess we should keep it as is.

I don't see a way either; closing for now.
